util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,181 @@
+    )
+    parser.add_argument(
+        "--db_name",
+        type=str,
+        default="gem5-vision",
+        help="Name of the database",
+    )
+    parser.add_argument(
+        "--collection_name",
+        type=str,
+        default="resources",
+        help="Name of the collection",
+    )
+    parser.add_argument(
+        "--auth_url",
+        default="https://realm.mongodb.com/api/client/v2.0/app/data-ejhjf/auth/providers/api-key/login",
+        type=str,
+        help="Authentication URL",
+    )
+    parser.add_argument(
+        "--db_url",
+        type=str,
+        default="https://data.mongodb-api.com/app/data-ejhjf/endpoint/data/v1",
+        help="Database URL",
+    )
+    parser.add_argument(
+        "--data_source",
+        type=str,
+        default="gem5-vision",
+        help="Data source name",
+    )
+    parser.add_argument(
+        "--output_dir",
+        type=str,
+        help="Output directory absolute path, default is the cwd. The resources.json and all resources will be saved in this directory",
+    )
+    args = parser.parse_args()
+
+    if args.output_dir == None:
+        output_path = pathlib.Path(pathlib.Path.cwd())
+    else:
+        output_path = pathlib.Path(args.output_dir)
+
+    token = get_token(args.auth_url, args.api_key)
+    resources = get_all_resources(
+        args.db_url,
+        args.data_source,
+        args.collection_name,
+        args.db_name,
+        token,
+    )
+    save_resources_to_file(
+        resources,
+        output_path,
+    )
+    download_resources(output_path)
+    return args
+
+
+def get_token(auth_url, api_key):
+    data = {"key": api_key}
+    data = json.dumps(data).encode("utf-8")
+    req = request.Request(
+        auth_url, data=data, headers={"content-type": "application/json"}
+    )
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        print(e)
+        print("Failed to get token")
+        exit(1)
+    token = json.loads(response.read().decode("utf-8"))["access_token"]
+    return token
+
+
+def get_all_resources(url, data_source, collection, database, token):
+    url = f"{url}/action/find"
+    data = {
+        "dataSource": data_source,
+        "collection": collection,
+        "database": database,
+    }
+
+    data = json.dumps(data).encode("utf-8")
+
+    headers = {
+        "Authorization": f"Bearer {token}",
+        "Content-Type": "application/json",
+    }
+
+    req = request.Request(url, data=data, headers=headers)
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        print("Failed to get resources")
+        print(e)
+        exit(1)
+    return json.loads(response.read().decode("utf-8"))["documents"]
+
+
+def save_resources_to_file(resources, output):
+    path = output.joinpath("resources.json")
+
+    with open(path, "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+def download_resources(output):
```

Comment:
Generally, types would be helpful. It would also be good to have docstrings for these functions.

Another point: Doesn't "download_resource.py" or something like that already have this function?

Reply:
I added typing and docstrings.

Yes, in `downloader.py` the `_download()` function does download. It is a more complex function and the majority of the above function will remain the same. The only change will be to call `_download()` function instead of `urlretrieve` as I will still need to parse which resources can be downloaded and get the filepath.

Reply:
I was actually referring to the `util/obtain-resource.py` script.

I'm ok with keeping this script as it is and not trying to combine with `util/obtain-resource.py`